packages/runtime/src/next/middleware/request-utils.ts

```diff
@@ -0,0 +1,193 @@
+      if (err instanceof MiddlewareError) throw err
+      const exception = new UnknownDraftFetchRequestError(
+        'Encountered unknown error while fetching draft cookies',
+      )
+      exception.cause = err
+      throw exception
+    })
+
+  const setCookieHeader = response.headers.getSetCookie()
+
+  if (setCookieHeader == null) {
+    throw new UnparseableDraftCookieResponseError(
+      "Received empty draft cookies. If you're seeing this error, please report a bug to https://github.com/makeswift/makeswift",
+    )
+  }
+
+  try {
+    return parseSetCookie(setCookieHeader)
+  } catch (err) {
+    const exception = new UnparseableDraftCookieResponseError(
+      "Could not parse draft cookies. If you're seeing this error, please report a bug to https://github.com/makeswift/makeswift",
+    )
+    exception.cause = err
+    throw exception
+  }
+}
+
+async function fetchDraftModeCookies(origin: string, draftSecret: string): Promise<Cookie[]> {
+  return await fetchCookies(new URL(CookieRequestEndpoints.DraftMode, origin), draftSecret)
+}
+
+async function fetchPreviewModeCookies(origin: string, draftSecret: string): Promise<Cookie[]> {
+  return await fetchCookies(new URL(CookieRequestEndpoints.PreviewMode, origin), draftSecret)
+}
+
+async function fetchDraftCookies(request: NextRequest, draftSecret: string): Promise<Cookie[]> {
+  if (isDraftModeRequest(request)) {
+    return await fetchDraftModeCookies(request.nextUrl.origin, draftSecret)
+  } else if (isPreviewModeRequest(request)) {
+    return await fetchPreviewModeCookies(request.nextUrl.origin, draftSecret)
+  }
+
+  return []
+}
+
+class MakeswiftDraftRequest extends NextRequest {
+  isMakeswiftDraftRequest = true as const
+}
+
+/**
+ * Creates a new `NextRequest` that can be used for fetching a working version
+ * of the page (that will also bypass the Vercel cache). Returns `null` if the
+ * given request does not have a draft secret that matches the `apiKey`.
+ */
+export async function createDraftRequest(
+  requestInit: NextRequest,
+  apiKey: string,
+): Promise<MakeswiftDraftRequest | null> {
+  const requestSecret = getDraftSecret(requestInit)
+
+  if (requestSecret == null || requestSecret !== apiKey) return null
+
+  const draftCookies = await fetchDraftCookies(requestInit, requestSecret)
+
+  // https://github.com/vercel/next.js/issues/52967#issuecomment-1644675602
+  // if we don't pass request twice, headers are stripped
+  const draftRequest = new MakeswiftDraftRequest(requestInit, requestInit)
+
+  draftCookies.forEach(({ name, value }) => {
+    draftRequest.cookies.set(name, value)
+  })
+
+  draftRequest.nextUrl.searchParams.delete(SearchParams.DraftMode)
+  draftRequest.nextUrl.searchParams.delete(SearchParams.PreviewMode)
+  draftRequest.headers.delete(HeaderNames.DraftMode)
+  draftRequest.headers.delete(HeaderNames.PreviewMode)
+
+  return draftRequest
+}
+
+export async function fetchDraftProxyResponse(
+  draftRequest: MakeswiftDraftRequest,
```

Comment:
If we're requiring a certain instance we should validate it at runtime as well. Remember not all consumers of our SDK use TypeScript